keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CleanerTests.kt

```diff
@@ -0,0 +1,158 @@
+  )
+
+  data class Fixture<T : DeliveryConfigRepository, R : ResourceRepository, A : ArtifactRepository>(
+    val deliveryConfigRepositoryProvider: (ResourceTypeIdentifier) -> T,
+    val resourceRepositoryProvider: () -> R,
+    val artifactRepositoryProvider: () -> A
+  ) {
+    private val clock: Clock = Clock.systemDefaultZone()
+    private val publisher: ApplicationEventPublisher = mockk(relaxUnitFun = true)
+
+    private val resourceTypeIdentifier: ResourceTypeIdentifier =
+      object : ResourceTypeIdentifier {
+        override fun identify(apiVersion: ApiVersion, kind: String): Class<out ResourceSpec> {
+          return when (kind) {
+            "security-group" -> DummyResourceSpec::class.java
+            "cluster" -> DummyResourceSpec::class.java
+            else -> error("unsupported kind $kind")
+          }
+        }
+      }
+
+    internal val deliveryConfigRepository: T = deliveryConfigRepositoryProvider(resourceTypeIdentifier)
+    internal val resourceRepository: R = resourceRepositoryProvider()
+    internal val artifactRepository: A = artifactRepositoryProvider()
+
+    val subject = Cleaner(
+      deliveryConfigRepository,
+      artifactRepository,
+      resourceRepository,
+      clock,
+      publisher
+    )
+
+    fun persist(deliveryConfig: DeliveryConfig) {
+      deliveryConfig.artifacts.forEach {
+        artifactRepository.register(it)
+      }
+      deliveryConfig.resources.forEach {
+        resourceRepository.store(it)
+      }
+      deliveryConfigRepository.store(deliveryConfig)
+    }
+  }
+
+  fun tests() = rootContext<Fixture<T, R, A>>() {
+    fixture {
+      Fixture(
+        deliveryConfigRepositoryProvider = this@CleanerTests::createDeliveryConfigRepository,
+        resourceRepositoryProvider = this@CleanerTests::createResourceRepository,
+        artifactRepositoryProvider = this@CleanerTests::createArtifactRepository
+      )
+    }
+
+    before {
+      persist(deliveryConfig)
+    }
+
+    after {
+      flush()
+    }
+
+    context("delivery config was created then deleted") {
+      before {
+        subject.delete(configName)
+      }
+      test("everything is deleted") {
+        expectThrows<NoSuchDeliveryConfigException> { deliveryConfigRepository.get(configName) }
+        expectThrows<NoSuchResourceException> { resourceRepository.get(firstResource.id) }
+        expectThat(artifactRepository.get(artifact.name, artifact.type, configName)).isEmpty()
+      }
+    }
+
+    context("delivery config is updated") {
+      context("artifact and resource have changed") {
+        before {
+          val updatedConfig = deliveryConfig.copy(
+            artifacts = setOf(newArtifact),
+            environments = setOf(firstEnv.copy(resources = setOf(secondResource)))
+          )
+          persist(updatedConfig)
+          subject.removeResources(old = deliveryConfig, new = updatedConfig)
+        }
+
+        test("no longer present resources are removed") {
```

Comment:
Add the opposite test to check that the new resources were _not_ inadvertently removed?